collector/collector.go

```diff
@@ -0,0 +1,607 @@
+						rns := make([]plugin.NamespaceElement, len(ns))
+						copy(rns, ns)
+
+						rns[indexOfDynamicElement+lengthOfNsPrefix].Value = strconv.Itoa(deviceID)
+						metric := plugin.Metric{
+							Timestamp: time.Now(),
+							Namespace: rns,
+							Data:      val,
+							Tags:      mt.Tags,
+							Config:    mt.Config,
+							Version:   PLUGIN_VERSION,
+						}
+						metrics = append(metrics, metric)
+					}
+				} else {
+					deviceID, err := strconv.Atoi(metricName[0])
+					if err != nil {
+						return nil, fmt.Errorf("In metric %s the given cpu id is invalid, err=%v", strings.Join(mt.Namespace.Strings(), "/"), err)
+					}
+					if deviceID > numOfDevices || deviceID < 0 {
+						return nil, fmt.Errorf("In metric %s the given cpu id is invalid, expected value in range 0-%d", strings.Join(mt.Namespace.Strings(), "/"), numOfDevices)
+					}
+
+					metric := plugin.Metric{
+						Timestamp: time.Now(),
+						Namespace: ns,
+						Data:      c.containers[rid].Stats.Cgroups.BlkioStats.IoMergedRecursive[deviceID],
+						Tags:      mt.Tags,
+						Config:    mt.Config,
+						Version:   PLUGIN_VERSION,
+					}
+					metrics = append(metrics, metric)
+				}
+			} // the end of switch statsType
+		} // the end of range over ids
+	}
+
+	if len(metrics) == 0 {
+		return nil, fmt.Errorf("No metrics found")
+	}
+
+	return metrics, nil
+}
+
+// GetMetricTypes returns list of available metrics
+func (c *collector) GetMetricTypes(cfg plugin.Config) ([]plugin.Metric, error) {
+	var err error
+	var metricTypes []plugin.Metric
+
+	// initialize containerData struct
+	data := container.ContainerData{
+		Stats: container.NewStatistics(),
+	}
+
+	dockerMetrics := []string{}
+	utils.FromCompositeObject(data, "", &dockerMetrics)
+	nscreator := nsCreator{dynamicElements: definedDynamicElements}
+	for _, metricName := range dockerMetrics {
+		ns := plugin.NewNamespace(PLUGIN_VENDOR, PLUGIN_NAME).
+			AddDynamicElement("docker_id", "an id of docker container")
+
+		if ns, err = nscreator.createMetricNamespace(ns, metricName); err != nil {
+			// skip this metric name which is not supported
+			log.WithFields(log.Fields{
+				"plugin": "docker_collector",
+				"block":  "GetMetricTypes",
+			}).Warnf("Error in creating metric %s: err=%s", metricName, err)
+			continue
+		}
+		metricType := plugin.Metric{
+			Namespace: ns,
+			Version:   PLUGIN_VERSION,
+		}
+		metricTypes = append(metricTypes, metricType)
+	}
+
+	return metricTypes, nil
+}
+
+// GetConfigPolicy returns plugin config policy
+func (c *collector) GetConfigPolicy() (plugin.ConfigPolicy, error) {
+	policy := plugin.NewConfigPolicy()
+	configKey := []string{"intel", "docker"}
+
+	policy.AddNewStringRule(configKey,
+		"endpoint",
+		false,
+		plugin.SetDefaultString("unix:///var/run/docker.sock"))
+
+	policy.AddNewStringRule(configKey,
+		"procfs",
+		false,
+		plugin.SetDefaultString("/proc"))
+
+	return *policy, nil
+}
+
+type collector struct {
```

Comment:
+1 to @kindermoumoute's comment because this is a type which is returned by `New()` in line 94, so it will be great to have a definition of this structure after const or variables (which are also unexported but are placed on the top despite this).